va/tlsalpn.go

```diff
@@ -58,36 +61,46 @@ func certAltNames(cert *x509.Certificate) []string {
 
 func (va *ValidationAuthorityImpl) tryGetChallengeCert(
 	ctx context.Context,
-	identifier identifier.ACMEIdentifier,
-	tlsConfig *tls.Config,
+	ident identifier.ACMEIdentifier,
 ) (*x509.Certificate, *tls.ConnectionState, core.ValidationRecord, error) {
-
-	allAddrs, resolvers, err := va.getAddrs(ctx, identifier.Value)
 	validationRecord := core.ValidationRecord{
-		DnsName:           identifier.Value,
-		AddressesResolved: allAddrs,
+		DnsName:           ident.Value,
+		AddressesResolved: []net.IP{},
 		Port:              strconv.Itoa(va.tlsPort),
-		ResolverAddrs:     resolvers,
+		ResolverAddrs:     bdns.ResolverAddrs{},
 	}
-	if err != nil {
-		return nil, nil, validationRecord, err
+
+	switch ident.Type {
+	case identifier.TypeDNS:
+		// Resolve IP addresses for the identifier
+		addrs, resolvers, err := va.getAddrs(ctx, ident.Value)
+		if err != nil {
+			return nil, nil, validationRecord, err
+		}
+		validationRecord.AddressesResolved, validationRecord.ResolverAddrs = addrs, resolvers
+	case identifier.TypeIP:
+		validationRecord.AddressesResolved = []net.IP{net.ParseIP(ident.Value)}
+	default:
+		// This should never happen. The calling function should check the
+		// identifier type.
+		return nil, nil, validationRecord, fmt.Errorf("Unknown identifier type: %s", ident.Type)
 	}
 
 	// Split the available addresses into v4 and v6 addresses
-	v4, v6 := availableAddresses(allAddrs)
+	v4, v6 := availableAddresses(validationRecord.AddressesResolved)
```

Comment:
As a general rule, don't pull values out of an object intended only for logging. Sure, it's a nice big blob that holds everything we might care about, but we never want our logic to have a dependency on our logging.